<a href="https://colab.research.google.com/github/deeclaudine/DM-/blob/main/lab02/Lab02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd

raw_data = {
    "order_id": [1001, 1002, 1003, 1004, 1005, 1006, 1007, 1008, 1009, 1010, 1011, 1012, 1013, 1014, 1015],
    "channel": ["site", "app", "app", "marketplace", "site", "app", "site", "marketplace", "app", "site", "marketplace", "app", "site", "site", "marketplace"],
    "order_sum": [1200, 2500, 1800, 3200, 900, 1500, 2800, 45000, 2100, 1700, 3600, 1100, 2400, 1900, 3300],
    "delivery_time": [25, np.nan, 40, 55, 20, 30, np.nan, 70, 35, 45, 50, 18, 27, np.nan, 60],
    "rating": [5, 4, 5, 3, 5, 4, 2, 1, 5, 4, 3, 5, 4, 5, 2]
}

df = pd.DataFrame(raw_data)
print(df)
df.to_csv("online_store_orders.csv", index=False)

    order_id      channel  order_sum  delivery_time  rating
0       1001         site       1200           25.0       5
1       1002          app       2500            NaN       4
2       1003          app       1800           40.0       5
3       1004  marketplace       3200           55.0       3
4       1005         site        900           20.0       5
5       1006          app       1500           30.0       4
6       1007         site       2800            NaN       2
7       1008  marketplace      45000           70.0       1
8       1009          app       2100           35.0       5
9       1010         site       1700           45.0       4
10      1011  marketplace       3600           50.0       3
11      1012          app       1100           18.0       5
12      1013         site       2400           27.0       4
13      1014         site       1900            NaN       5
14      1015  marketplace       3300           60.0       2


Для чего используется параметр index=False? Что произойдёт при сохранении таблицы без этого параметра?

Ответ: Параметр index=False запрещает pandas записывать индексы строк (0, 1, 2...) в CSV-файл как отдельный безымянный столбец. Если его не указать, при следующей загрузке файла появится лишний столбец Unnamed: 0, что нарушит структуру данных и усложнит дальнейший анализ.

In [2]:
store_df = pd.read_csv("online_store_orders.csv")

print("Первые 5 строк:")
print(store_df.head())

print("\nПоследние 5 строк:")
print(store_df.tail())

print("\nРазмер таблицы:")
print(store_df.shape)

print("\nНазвания столбцов:")
print(store_df.columns.tolist())

print("\nИнформация о таблице:")
store_df.info()

Первые 5 строк:
   order_id      channel  order_sum  delivery_time  rating
0      1001         site       1200           25.0       5
1      1002          app       2500            NaN       4
2      1003          app       1800           40.0       5
3      1004  marketplace       3200           55.0       3
4      1005         site        900           20.0       5

Последние 5 строк:
    order_id      channel  order_sum  delivery_time  rating
10      1011  marketplace       3600           50.0       3
11      1012          app       1100           18.0       5
12      1013         site       2400           27.0       4
13      1014         site       1900            NaN       5
14      1015  marketplace       3300           60.0       2

Размер таблицы:
(15, 5)

Названия столбцов:
['order_id', 'channel', 'order_sum', 'delivery_time', 'rating']

Информация о таблице:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column     

Количество строк: 15
Количество столбцов: 5
Числовые столбцы: order_id, order_sum, delivery_time, rating
Столбцы с пропущенными значениями: delivery_time
Количество пропущенных значений: 3

In [7]:
# 5.1 Выбор столбцов
print("Только сумма и время доставки:")
print(store_df[["order_sum", "delivery_time"]])

# 5.2 Выбор строк с индексами от 3 до 7 включительно
slice_df = store_df.iloc[3:8, [2, 3]]
print("\nСрез с индексами 3-7 (order_sum и delivery_time):")
print(slice_df)

# 5.3 Фильтрация
premium_orders = store_df[(store_df["order_sum"] > 2000) & (store_df["rating"] == 5)]
print("\nПремиум заказы (сумма > 2000 и рейтинг == 5):")
print(premium_orders)

Только сумма и время доставки:
    order_sum  delivery_time
0        1200           25.0
1        2500            NaN
2        1800           40.0
3        3200           55.0
4         900           20.0
5        1500           30.0
6        2800            NaN
7       45000           70.0
8        2100           35.0
9        1700           45.0
10       3600           50.0
11       1100           18.0
12       2400           27.0
13       1900            NaN
14       3300           60.0

Срез с индексами 3-7 (order_sum и delivery_time):
   order_sum  delivery_time
3       3200           55.0
4        900           20.0
5       1500           30.0
6       2800            NaN
7      45000           70.0

Премиум заказы (сумма > 2000 и рейтинг == 5):
   order_id channel  order_sum  delivery_time  rating  is_high_value
8      1009     app       2100           35.0       5          False


Почему для объединения условий используется &, а не and?

Ответ: Оператор & является побитовым (векторизованным) и применяется к каждому элементу столбца (Series) по отдельности. Ключевое слово and в Python пытается оценить истинность всего объекта целиком, что для pandas Series вызывает ошибку ValueError, так как неясно, как интерпретировать истинность массива данных.

In [4]:
# 6.1 Пропущенные значения
print("Пропущенные значения в каждом столбце:")
print(store_df.isna().sum())

# 6.2 Описательная статистика
print("\nОписательная статистика:")
print(store_df.describe())

# 6.3 Среднее и медиана времени доставки
mean_delivery = store_df["delivery_time"].mean()
median_delivery = store_df["delivery_time"].median()
print(f"\nСреднее время доставки: {mean_delivery:.2f}")
print(f"Медианное время доставки: {median_delivery:.2f}")

# 6.4 Распределение рейтингов
print("\nРаспределение рейтингов:")
print(store_df["rating"].value_counts())

Пропущенные значения в каждом столбце:
order_id         0
channel          0
order_sum        0
delivery_time    3
rating           0
dtype: int64

Описательная статистика:
          order_id    order_sum  delivery_time     rating
count    15.000000     15.00000      12.000000  15.000000
mean   1008.000000   5000.00000      39.583333   3.800000
std       4.472136  11096.33144      16.654010   1.320173
min    1001.000000    900.00000      18.000000   1.000000
25%    1004.500000   1600.00000      26.500000   3.000000
50%    1008.000000   2100.00000      37.500000   4.000000
75%    1011.500000   3000.00000      51.250000   5.000000
max    1015.000000  45000.00000      70.000000   5.000000

Среднее время доставки: 39.58
Медианное время доставки: 37.50

Распределение рейтингов:
rating
5    6
4    4
3    2
2    2
1    1
Name: count, dtype: int64


1. Сколько заказов не имеют информации о времени доставки?
3 заказа.
2. Какой показатель имеет значение max, существенно отличающееся от остальных?
order_sum (максимум 45 000, что сильно выделяется на фоне остальных сумм ~1000–3000).
3. Почему среднее значение может отличаться от медианы?
 Среднее арифметическое чувствительно к выбросам (аномально высоким или низким значениям), которые "тянут" его в свою сторону. Медиана же показывает центральное значение упорядоченного ряда и устойчива к выбросам, поэтому лучше отражает "типичное" значение.
4. Какая оценка встречается наиболее часто? Оценка 5 (встречается 6 раз).

In [5]:
# 7.1 Создание производного признака
store_df["is_high_value"] = store_df["order_sum"] > 3000
print("Проверка нового столбца:")
print(store_df[["order_id", "order_sum", "is_high_value"]])

# 7.2 Группировка по каналу продаж
channel_stats = store_df.groupby("channel")["order_sum"].mean().reset_index()
print("\nСредняя сумма заказа по каналам:")
print(channel_stats)

# 7.3 Сортировка результата
channel_stats = channel_stats.sort_values(by="order_sum", ascending=False)
print("\nОтсортированные каналы по убыванию средней суммы:")
print(channel_stats)

Проверка нового столбца:
    order_id  order_sum  is_high_value
0       1001       1200          False
1       1002       2500          False
2       1003       1800          False
3       1004       3200           True
4       1005        900          False
5       1006       1500          False
6       1007       2800          False
7       1008      45000           True
8       1009       2100          False
9       1010       1700          False
10      1011       3600           True
11      1012       1100          False
12      1013       2400          False
13      1014       1900          False
14      1015       3300           True

Средняя сумма заказа по каналам:
       channel     order_sum
0          app   1800.000000
1  marketplace  13775.000000
2         site   1816.666667

Отсортированные каналы по убыванию средней суммы:
       channel     order_sum
1  marketplace  13775.000000
2         site   1816.666667
0          app   1800.000000


Средняя сумма заказа по каналам: marketplace (~13 775 руб.), site (~1 967 руб.), app (~1 800 руб.).

Количество заказов с is_high_value = True: 4 заказа.

Количество пропущенных значений: 3 (в столбце delivery_time).

Влияние выброса: Заказ на 45 000 руб. в канале "marketplace" искусственно завышает среднюю сумму по этому каналу и в целом по магазину.

Различие среднего и медианного времени доставки невелико (39.58 против 37.5), что говорит об относительно симметричном распределении времени доставки, несмотря на отдельные задержки.

Итоговый аналитический вывод

В наборе данных содержится 15 заказов. В столбце delivery_time обнаружено 3 пропущенных значения, требующих уточнения. Среднее время доставки составляет 39.6 часов, медианное — 37.5 часов, что указывает на стабильность логистики. Средняя сумма заказа по каналам продаж составляет: marketplace — 13 775 руб., site — 1 967 руб., app — 1 800 руб. Выявлено 4 заказа с высокой стоимостью (более 3000 руб.). При интерпретации средней суммы заказа необходимо учитывать наличие аномального выброса (заказ на 45 000 руб.), который искажает среднее значение в канале marketplace; для более точной оценки "типичного" чека рекомендуется использовать медиану или анализировать данные без учета экстремальных значений.